kani-driver/src/harness_runner.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use anyhow::Result;
+use kani_metadata::HarnessMetadata;
+use std::path::{Path, PathBuf};
+
+use crate::call_cbmc::VerificationStatus;
+use crate::session::KaniSession;
+use crate::util::specialized_harness_name;
+
+pub(crate) struct HarnessRunnerSession<'sess> {
+    /// The underlying kani session
+    pub sess: &'sess KaniSession,
+    /// The build CBMC goto binary for the "whole program" (will be specialized to each proof harness)
+    pub linked_obj: &'sess Path,
+    /// The directory we should output cbmc-viewer reports to
+    pub report_base: &'sess Path,
+    /// A behavior difference between `kani` and `cargo kani`: `cargo kani` never deleted the specialized goto binaries
+    pub retain_specialized_harnesses: bool,
+
+    /// The collection of symtabs that went into the goto binary
+    /// (TODO: this is only for --gen-c, which possibly should not be done here?)
+    pub symtabs: &'sess [PathBuf],
+}
+
+pub(crate) struct HarnessResult<'sess> {
+    pub harness: &'sess HarnessMetadata,
+    pub result: VerificationStatus,
+}
+
+#[derive(Default)]
+pub(crate) struct HarnessResults<'sess> {
+    pub successes: Vec<HarnessResult<'sess>>,
+    pub failures: Vec<HarnessResult<'sess>>,
+}
+
+impl<'sess> HarnessRunnerSession<'sess> {
+    pub(crate) fn check_all_harnesses<'a>(
+        &self,
+        harnesses: &'a [HarnessMetadata],
+    ) -> Result<HarnessResults<'a>> {
+        let sorted_harnesses = crate::metadata::sort_harnesses_by_loc(harnesses);
+
+        let mut results = HarnessResults::default();
+
+        for harness in &sorted_harnesses {
+            let harness_filename = harness.pretty_name.replace("::", "-");
+            let report_dir = self.report_base.join(format!("report-{}", harness_filename));
+            let specialized_obj = specialized_harness_name(self.linked_obj, &harness_filename);
+            if !self.retain_specialized_harnesses {
+                let mut temps = self.sess.temporaries.borrow_mut();
+                temps.push(specialized_obj.to_owned());
+            }
+            self.sess.run_goto_instrument(
+                self.linked_obj,
+                &specialized_obj,
+                self.symtabs,
+                &harness.mangled_name,
+            )?;
+
+            let result = self.sess.check_harness(&specialized_obj, &report_dir, harness)?;
+            let wrapped_result = HarnessResult { harness, result };
+            if wrapped_result.result == VerificationStatus::Failure {
+                results.failures.push(wrapped_result);
+            } else {
+                results.successes.push(wrapped_result);
+            }
+        }
+
+        Ok(results)
+    }
+}
+
+impl KaniSession {
+    pub(crate) fn check_harness(
```

Comment:
Shouldn't this be part of the new HarnessRunner?

Reply:
Probably, but there's also going to be some other changes coming here eventually.

I'm not sure I'll get to this as part of the initial parallel runner, but I will eventually "lift" all the output code outwards, so that we don't have e.g. interleaving problems, and we can more deeply inspect the actual results. (For example, I intend `VerificationStatus::Failure` to start carrying information about why it failed. Both for "assess" and to lift the "printing" out of the "running" to eventually better solve the interleaving output problem, which I might just initially solve with an annoying mutex.)

Reply:
Interesting... how are you planning to display the verification status for the user? I'm thinking here about the case where there are 2 harnesses stuck due to unlimited unwinding. We won't be able to print CBMC messages on the fly anymore, right?

Reply:
Right. When running in parallel we wouldn't be able to use that approach to display unwinding status anymore.

Parallel running will initially be an unstable feature that will require `--enable-unstable`. Making it work well will take a lot more work (that's frankly already necessary: see issues complaining about cbmc's output, and the need for better communication to users about unbounded unwinding.)